File: _inbox/PDI/01-orquestracao-corporativa-n8n/atividade-3-payloads-pesados-nos-customizados/3-lib/payload-lib.py
# ============================================================
# payload_lib.py — Biblioteca reutilizável de nós Code Python
# PDI A3: Nós customizados / expressões avançadas
# Apenas stdlib — sem dependências externas.
# ============================================================
import json
from collections import Counter, defaultdict
from itertools import islice

# ------------------------------------------------------------
# 1) Chunking streaming (itertools.islice)
# ------------------------------------------------------------
def chunk(iterable, size=100):
    it = iter(iterable)
    while True:
        batch = list(islice(it, size))
        if not batch:
            break
        yield batch

# ------------------------------------------------------------
# 2) Dedupe por chave primitiva (O(n))
# ------------------------------------------------------------
def dedupe(rows, key='id'):
    seen = set()
    result = []
    for row in rows:
        k = row.get(key)
        if k in seen:
            continue
        seen.add(k)
        result.append(row)
    return result

# ------------------------------------------------------------
# 3) Agregação com Counter/defaultdict (O(n), nunca O(n²))
# ------------------------------------------------------------
def aggregate(rows, key='tipo', value=None):
    counts = Counter(r.get(key) for r in rows)
    totals = defaultdict(int)
    if value:
        for r in rows:
            try:
                totals[r.get(key)] += r.get(value) or 0
            except TypeError:
                pass
    return {
        'counts': dict(counts),
        'totals': dict(totals),
    }

# ------------------------------------------------------------
# 4) Decodifica payload 1x (vem como string)
# ------------------------------------------------------------
def parse_payload(payload):
    if isinstance(payload, str):
        return json.loads(payload)
    if isinstance(payload, dict) and 'rows' in payload and isinstance(payload['rows'], str):
        payload['rows'] = json.loads(payload['rows'])
    return payload

# ------------------------------------------------------------
# 5) Saída para n8n
# ------------------------------------------------------------
def to_output(data):
    return [{'json': data}]

# ------------------------------------------------------------
# Exemplo de uso dentro de um nó Code Python:
#
#   data = parse_payload(_input[0]['json'])
#   rows = dedupe(data.get('rows', []), key='id')
#   agg = aggregate(rows, key='tipo')
#   return to_output({'count': len(rows), 'aggregation': agg})
# ------------------------------------------------------------

File: _inbox/PDI/01-orquestracao-corporativa-n8n/atividade-3-payloads-pesados-nos-customizados/3-lib/payload-lib.js
// ============================================================
// payload-lib.js — Biblioteca reutilizável de nós Code JS
// PDI A3: Nós customizados / expressões avançadas
// Use como referência para colar (copiar) dentro de nós Code.
// ============================================================

// ------------------------------------------------------------
// 1) Chunking streaming — processa em lotes sem carregar tudo
// ------------------------------------------------------------
export function* chunk(items, size = 100) {
  for (let i = 0; i < items.length; i += size) {
    yield items.slice(i, i + size);
  }
}

// ------------------------------------------------------------
// 2) Normalização em UMA passada (O(n))
//    Filtra + dedupe + transforma no mesmo loop.
// ------------------------------------------------------------
export function normalizeStream(items, opts = {}) {
  const { dedupeKey = 'id', map = (j) => j } = opts;
  const seen = new Set();
  const out = [];

  for (const item of items) {
    const j = item.json || item;
    if (opts.filter && !opts.filter(j)) continue;

    const key = j[dedupeKey];
    if (key !== undefined) {
      if (seen.has(key)) continue;
      seen.add(key);
    }

    out.push({ json: map(j) });
  }
  return out;
}

// ------------------------------------------------------------
// 3) Dedupe por chave primitiva (nunca por objeto)
// ------------------------------------------------------------
export function dedupe(items, keyFn) {
  const seen = new Set();
  return items.filter((item) => {
    const key = keyFn(item.json);
    if (seen.has(key)) return false;
    seen.add(key);
    return true;
  });
}

// ------------------------------------------------------------
// 4) Agregação com Map (O(n), evita busca por item)
// ------------------------------------------------------------
export function aggregate(items, keyFn, reduce) {
  const map = new Map();
  for (const item of items) {
    const key = keyFn(item.json);
    if (!map.has(key)) map.set(key, { key, count: 0, acc: undefined });
    const entry = map.get(key);
    entry.count += 1;
    entry.acc = reduce ? reduce(entry.acc, item.json) : entry.acc;
  }
  return [...map.values()];
}

// ------------------------------------------------------------
// 5) Memoização entre execuções ($getWorkflowStaticData)
// ------------------------------------------------------------
export function memoizeGlobal(key, computeFn) {
  const data = $getWorkflowStaticData('global');
  if (data[key] === undefined) data[key] = computeFn();
  return data[key];
}

// ------------------------------------------------------------
// 6) Saída para n8n (garante formato { json })
// ------------------------------------------------------------
export function toOutput(rows) {
  return rows.map((json) => ({ json }));
}

// ------------------------------------------------------------
// Exemplo de uso dentro de um nó Code:
//
//   const { chunk, normalizeStream, toOutput } = ...
//   const items = $input.all();
//   const normalized = normalizeStream(items, {
//     dedupeKey: 'id',
//     filter: (j) => j.score >= 70,
//     map: (j) => ({ id: j.id, name: String(j.name).toUpperCase() }),
//   });
//   return toOutput(normalized);
// ------------------------------------------------------------